metronome/hotstuff/service/test/src/io/iohk/metronome/hotstuff/service/sync/ViewSynchronizerProps.scala

```diff
@@ -0,0 +1,374 @@
+
+    val retryTimeout = 5.seconds
+    val syncTimeout  = fixture.rounds * retryTimeout * 2
+    val synchronizer = new ViewSynchronizer[Task, TestAgreement](
+      federation = fixture.federation,
+      getStatus = fixture.getStatus,
+      retryTimeout = retryTimeout
+    )
+
+    val test = for {
+      status <- synchronizer.sync.timeout(syncTimeout).attempt
+      events <- fixture.syncEventsRef.get
+
+      quorumSize = fixture.federation.quorumSize
+
+      indexOfQuorum = fixture.responses.indexWhere { responseMap =>
+        responseMap.values.collect { case TestResponse.ValidStatus(_) =>
+        }.size >= quorumSize
+      }
+      hasQuorum = indexOfQuorum >= 0
+
+      invalidResponseCount = {
+        val responses =
+          if (hasQuorum) fixture.responses.take(indexOfQuorum + 1)
+          else fixture.responses
+        responses
+          .flatMap(_.values)
+          .collect { case TestResponse.InvalidStatus(_) =>
+          }
+          .size
+      }
+
+      invalidEventCount = {
+        events.collect { case x: SyncEvent.InvalidStatus[_] =>
+        }.size
+      }
+
+      pollSizes = events.collect { case SyncEvent.StatusPoll(statuses) =>
+        statuses.size
+      }
+
+      responseCounter <- fixture.responseCounterRef.get
+    } yield {
+      val statusProps = status match {
+        case Right(status) =>
+          "status" |: all(
+            "quorum" |: hasQuorum,
+            "reports polls each round" |:
+              pollSizes.size == indexOfQuorum + 1,
+            "stop at the first quorum" |:
+              pollSizes.last >= quorumSize &&
+              pollSizes.init.forall(_ < quorumSize),
+            "reports all invalid" |:
+              invalidEventCount == invalidResponseCount
+          )
+
+        case Left(ex: TimeoutException) =>
+          "timeout" |: all(
+            "no quorum" |: !hasQuorum,
+            "empty polls" |: pollSizes.forall(_ < quorumSize),
+            "keeps polling" |: pollSizes.size >= fixture.rounds,
+            "reports all invalid" |: invalidEventCount == invalidResponseCount
+          )
+
+        case Left(ex) =>
+          ex.getMessage |: false
+      }
+
+      all(
+        statusProps,
+        "poll everyone in each round" |:
+          responseCounter.values.toList.distinct.size <= 2 // Some members can get an extra query, down to timing.
+      )
+    }
+
+    val testFuture = test.runToFuture
+
+    scheduler.tick(syncTimeout)
+
+    testFuture.value.get.get
+  }
+
+  property("median") = forAllNoShrink(
+    for {
+      m   <- arbitrary[Int].map(_.toLong)
+      l   <- Gen.posNum[Int]
+      h   <- Gen.oneOf(l, l + 1)
+      ls  <- Gen.listOfN(l, Gen.posNum[Int].map(m - _))
+      hs  <- Gen.listOfN(l, Gen.posNum[Int].map(m + _))
+      rnd <- arbitrary[Int].map(new Random(_))
+    } yield (m, rnd.shuffle(ls ++ Seq(m) ++ hs))
+  ) { case (m, xs) =>
+    m == ViewSynchronizer.median(NonEmptySeq.fromSeqUnsafe(xs))
```

Comment:
I think you meant to generate `hs` using `h`, otherwise we're only testing lists of odd lengths. This would fail however for `h == l + 1`. We need `h ∈ [l - 1, l]` for the current impl.

Reply:
Thanks, you have eagle eyes!

Reply:
Done, good catch.